### PR TITLE
Add `.numeric()` to `ZodString` 

### PR DESCRIPTION
**Pull request by icecreamparlor**
Hi 👋

I think it would be great to add `numericString` in ZodString.
It is just util for `regex` checker, but I belive it would be useful in many cases.

Related Issue : https://github.com/colinhacks/zod/issues/1240

### CODE REVIEW

File: src/__tests__/string.test.ts [hunk 1]
```diff
@@ -262,3 +265,25 @@ test("datetime parsing", () => {
...
+  expect(() => numericString.parse("1")).not.toThrow();
+  expect(() =>
+    numericString.parse(
+      "99999000000000000000000000000000000000000000000000000000"
+    )
+  ).not.toThrow();
+  expect(() => numericString.parse("3.141592")).not.toThrow();
+  expect(() => numericString.parse("-1")).not.toThrow();
+  expect(() => numericString.parse("-999.1231234134"));
+});
```
Comment (djjavo) [inline src/__tests__/string.test.ts:289]: A very minor (test) suggestion: as this `numericString` check appears to work for barcodes (i.e. a numeric string which is often zero padded, such as `0287730000000`), would it be worth adding a sample barcode to a test?

` expect(() => numericString.parse("0287730000000")).not.toThrow();`

Comment (icecreamparlor) [reply src/__tests__/string.test.ts:290]: Sounds great. I added test case. Thanks for suggestion : )

**Comment by igalklebanov** [comment]
`Numeric` seems redundant given its called in `z.string()` context. Why not simply `z.string().numeric()`?

File: src/types.ts [hunk 2]
```diff
@@ -739,6 +739,14 @@ export class ZodString extends ZodType<string, ZodStringDef> {
       ...errorUtil.errToObj(options?.message),
     });
   }
+  
+  numericString(message?: errorUtil.ErrMessage) {
+    return this._addCheck({
+      kind: "regex",
```
Comment (finaloop-igal) [inline src/types.ts:745]: Should align with other regex QoL methods and add its own kind `"numeric"` and `isNumeric()` method.

Comment (igalklebanov) [inline src/types.ts:745]: I think it should have a kind `"numeric"` and `isNumeric()` method to align with other regex QoL checks.

Comment (icecreamparlor) [reply src/types.ts]: Okay. I added `"numeric"` kind and `isNumeric()`.

**Comment by icecreamparlor** [comment]
> Numeric seems redundant given its called in z.string() context. Why not simply z.string().numeric()?

I think `numeric` seems better. I renamed `numericString` to `numeric`

**Review by dleo9307 — APPROVED** [review]
great PR, we really need it!

File: src/__tests__/string.test.ts [hunk 3]
```diff
@@ -262,3 +275,27 @@ test("datetime parsing", () => {
...
+
+  expect(() => numeric.parse("1")).not.toThrow();
+  expect(() =>
+    numeric.parse(
+      "99999000000000000000000000000000000000000000000000000000"
+    )
+  ).not.toThrow();
+  expect(() => numeric.parse("3.141592")).not.toThrow();
+  expect(() => numeric.parse("-1")).not.toThrow();
+  expect(() => numeric.parse("-999.1231234134"));
```
Comment (dleo9307) [inline src/__tests__/string.test.ts:299]:
```suggestion
  expect(() => numeric.parse("-999.1231234134")).not.toThrow();
```

**Comment by icecreamparlor** [comment]
(cc. @colinhacks)

**Comment by einarpersson** [comment]
Great, was just looking for this! Hope it gets merged soon!

**Comment by icecreamparlor** [comment]
Any updates ..? 
I think this PR can help many people : )

**Comment by jacksonblankenship** [comment]
I could really use this for one of my current projects 👀

**Comment by softcam-claiton** [comment]
Hello, any updates on when this will be merged?

**Comment by colinhacks** [comment]
Thanks for the excellent PR here, and apologies for taking so long to address it. At the moment, I'm trying to keep Zod's core feature set as small as possible; since this is achievable with `.regex()` or `.refine()` I'm inclined to keep it separate. 

Zod 4 will expose a bunch of utility regexes, and I'll be sure this one is included.

**Pull request closed without merging**